nservicebus/messaging/message-type-detection.md

```diff
@@ -1,24 +1,24 @@
 ---
 title: Message type detection
 summary: Describes the rules for mapping a transport message to a message type
-reviewed: 2021-06-05
+reviewed: 2023-08-04
 component: Core
+related:
+- samples/consumer-driven-contracts
 ---
 
-In order to invoke the correct message handlers for incoming messages, NServiceBus must be able to map the incoming transport message to a [message type](/nservicebus/messaging/messages-events-commands.md).
+NServiceBus sets the [`NServiceBus.EnclosedMessageTypes` header](/nservicebus/messaging/headers.md#serialization-headers-nservicebus-enclosedmessagetypes) to the [AssemblyQualifiedName](https://docs.microsoft.com/en-us/dotnet/api/system.type.assemblyqualifiedname) of the emitted message type when sending, publishing or replying.
+
+It is possible to change the enclosed message type header for all emitted messages by using a behavior. The [consumder-driven-contracts sample](/samples/consumer-driven-contracts) shows how to manipulate the header for outgoing messages.
+
+To invoke the correct message handlers for incoming messages, NServiceBus must be able to map the incoming transport message to a [message type](/nservicebus/messaging/messages-events-commands.md).
 
 The mapping rules are as follows:
 
 1. If the message contains the [`NServiceBus.EnclosedMessageTypes` header](/nservicebus/messaging/headers.md#serialization-headers-nservicebus-enclosedmessagetypes), the header value is used to find the message type. The header value may contain:
+   - The [AssemblyQualifiedName](https://docs.microsoft.com/en-us/dotnet/api/system.type.assemblyqualifiedname) of the message type (with or without private key are both supported). In cases when the assembly qualified type is not known by the endpoint NServiceBus will fall back using the [FullName](https://docs.microsoft.com/en-us/dotnet/api/system.type.fullname) of the message type.
    - The [FullName](https://docs.microsoft.com/en-us/dotnet/api/system.type.fullname) of the message type.
-   - The [AssemblyQualifiedName](https://docs.microsoft.com/en-us/dotnet/api/system.type.assemblyqualifiedname) of the message type (with or without private key are both supported)
    - The Name of the type, without the assembly name.
```

Comment:
what's the difference to full name? Is this assuming that there is no namespace either?

Reply:
I have no idea.

Reply:
At first glance, I couldn't find a usage in the relevant paths that goes for `.Name`. But it looks like for types that are in the default namespace `FullName` is equal to `Name` and I guess if there is no ambiguity the type can get loaded. I wonder if we have a test for that

Reply:
Maybe @mikeminutillo @HEskandari or @WojcikMike know since it was added in https://github.com/Particular/docs.particular.net/pull/4949/commits/a82a71719fa2d68e5ec50597d8bfc80714aadeef

Reply:
I believe Hadi was the main driver behind this change. I remember reviewing it but I don't see a test for a message type in the default namespace. The idea that [two types might match and we might pick the wrong one was brought up](https://github.com/Particular/NServiceBus/pull/5696/files#r461269356) which is why there is a debug log statement for the selected type. I don't remember any more details than that, sorry.